src/nvim/lua/treesitter.c

```diff
@@ -1375,13 +1379,17 @@ static int query_next_match(lua_State *L)
   TSQueryMatch match;
   if (ts_query_cursor_next_match(cursor, &match)) {
     lua_pushinteger(L, match.pattern_index + 1);  // [index]
+    // match: TSMatch = { [integer]: TSNode }
```

Comment:
Note for myself: I'm now thinking about a new idea of adding an explicit method to the table `TSMatch` (owned by C code):

`match:nodes(capture_id) -> TSNode[]`, returns the *list* of captured nodes (#24738) for a capture index.

I'm not yet sure this would be the best design, but some benefit of this would be:

- We can have backward compatibility without a breaking change in #24738; `match[integer]` can still return the last TSNode (eventually this can be deprecated).
- `set_match()` can be executed lazily: the captured nodes entries could be filled in only when requested, possibly leading to some performance gain. Many of the captures would not have a predicate, in which `match:nodes(...)` or `match[integer]` won't be called.
- Even if matches are alternating, we don't have to call `set_match()` again.

Note:

- Might need to define `__pair` in the metatable to make `for capture_id, node in pairs(match)` work correctly and same as before without breaking change.
- Do not rely on object identity (i.e. whether are same table instance) to test identity of a match. We can add a new field like `match.id`.

Reply:
> Might need to define __pairs in the metatable to make for capture_id, node in pairs(match) work correctly and same as before without breaking change.

Not in Lua 5.1 😉